File: Pyhton3 Implementations/Trees/BinaryTree/BinaryTree.py
class BinaryTree:
    def __init__(self):
        self.root = None

    class Node:
        def __init__(self, value):
            self.value = value
            self.left = None
            self.right = None

    def populate(self):
        value = int(input("Enter the value for root Node:"))
        self.root = self.Node(value)
        self.populate_helper(self.root)

    def populate_helper(self, node):
        left = input(f"Enter the left value for {node.value}:")
        if left!="none":
            node.left = self.Node(left)
            self.populate_helper(node.left)

        right = input(f"Enter the right value for {node.value}:")
        if right !="none":
            node.right = self.Node(right)
            self.populate_helper(node.right)

    def display(self):
        self.display_helper(self.root, 0)

    def display_helper(self, node, level):
        if node is None:
            return

        self.display_helper(node.right, level + 1)

        if level != 0:
            for i in range(level - 1):
                print("|        ", end="")
            print("|------->" + str(node.value))
        else:
            print(node.value)
        
        self.display_helper(node.left, level + 1)

    def in_order_traversal(self):
        result = []
        self._in_order_recursive(self.root, result)
        print("IN ORDER TRAVERSAL")
        print(result) 
    
    def _in_order_recursive(self, current_node, result):
        if current_node:
            self._in_order_recursive(current_node.left, result)
            result.append(current_node.value)
            self._in_order_recursive(current_node.right, result)


    def pre_order_traversal(self):
        result = []
        self._pre_order_recursive(self.root, result)
        print("PRE ORDER TRAVERSAL")
        print(result) 
    
    def _pre_order_recursive(self, current_node, result):
        if current_node:
            result.append(current_node.value)
            self._pre_order_recursive(current_node.left, result)
            self._pre_order_recursive(current_node.right, result)


    def post_order_traversal(self):
        result = []
        self._post_order_recursive(self.root, result)
        print("POST ORDER TRAVERSAL")
        print(result) 
    
    def _post_order_recursive(self, current_node, result):
        if current_node:
            self._post_order_recursive(current_node.left, result)
            self._post_order_recursive(current_node.right, result)
            result.append(current_node.value)
        
    def level_order_traversal(self):
        result = []
        root=self.root
        self.level_order_traversal_(root, result)
        print("LEVEL ORDER TRAVERSAL")
        print(result) 

    def level_order_traversal_(self,root,result):
        if root is None:
            return
        
        queue = [root]  

        while queue:
            node = queue.pop(0) 

            result.append(node.value)  

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        return result

    # def find_min_max(self):
    #     return self.find_min_max_rec(self.root)
    
  
    # def find_min_max_rec(self, root):
    #     if root is None:
    #         return float('inf'), float('-inf')

    #     left_min, left_max = self.find_min_max_rec(root.left)
    #     right_min, right_max = self.find_min_max_rec(root.right)

    #     min_val = min(root.value, left_min, right_min)
    #     max_val = max(root.value, left_max, right_max)

    #     return min_val, max_val




# Example usage:
if __name__ == "__main__":
    tree = BinaryTree()
    tree.populate()
    tree.display()
    tree.in_order_traversal()
    tree.post_order_traversal()
    tree.pre_order_traversal()
    tree.level_order_traversal()

File: JAVA Implementations/src/Trees/BinaryTree.java
package Trees;

import java.util.Scanner;

public class BinaryTree {
    public BinaryTree(){}
    private static class Node{
        private final int  value;
        private Node left;
        private Node right;
        public Node(int value){
            this.value  = value;
        }
    }
    private Node root = null;
    public void populate(Scanner scanner){
        System.out.println("Enter the value for root Node");
        int value = scanner.nextInt();
        root = new Node(value);
        populate(scanner,root);
    }
    public void populate(Scanner scanner,Node node){
        System.out.println("Do you want to enter left of " + node.value);
        boolean left = scanner.nextBoolean();
        if(left){
            System.out.println("Enter the value of the left of " + node.value);
            int value = scanner.nextInt();
            node.left = new Node(value);
            populate(scanner,node.left);

        }
        System.out.println("Do you want to enter right of " + node.value);
        boolean right = scanner.nextBoolean();
        if(right){
            System.out.println("Enter the value of the right of " + node.value);
            int value = scanner.nextInt();
            node.right = new Node(value);
            populate(scanner,node.right);
        }
    }
    public void preOrder(){
        preOrder(root);
        System.out.println("END");
    }
    private void preOrder(Node node){
        if(node == null){
            return;
        }
        System.out.print(node.value+"->");
        preOrder(node.left);
        preOrder(node.right);
    }

    public void inOrder(){
        inOrder(root);
        System.out.println("END");
    }
    private void inOrder(Node node){
        if(node == null){
            return;
        }
        inOrder(node.left);
        System.out.print(node.value+"->");
        inOrder(node.right);
    }
    public void postOrder(){
        postOrder(root);
        System.out.println("END");
    }
    private void postOrder(Node node){
        if(node == null){
            return;
        }
        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.value+"->");

    }

    public void display() {
        display(root, 0);
    }

    private void display(Node node, int level) {
        if (node == null) {
            return;
        }
        display(node.right, level + 1);

        if (level != 0) {
            for (int i = 0; i < level - 1; i++) {
                System.out.print("|\t\t");
            }
            System.out.println("|------->" + node.value);
        } else {
            System.out.println(node.value);
        }
        display(node.left, level + 1);
    }

}
